components/org.wso2.carbon.identity.tag.manager.core/src/main/java/org/wso2/carbon/identity/tag/manager/core/internal/impl/TagManagementServiceImpl.java

```diff
@@ -0,0 +1,271 @@
+        if (!isTagNameValid(tag.getName())) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_NAME.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_NAME.getMessage());
+        }
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        // Check if tag already existing.
+        if (tagManagementDAO.loadTag(tag.getName(), tag.getType(), tenantUuid) != null) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_TAG_ALREADY_EXISTS.getCode(),
+                    String.format(ErrorMessage.ERROR_CODE_TAG_ALREADY_EXISTS.getMessage(), tag.getName(),
+                            tenantUuid));
+        }
+        String tagUuid = tagManagementDAO.storeTag(tag, tenantUuid);
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Provided JSON request content is not in the valid format:");
+        }
+        return tagUuid;
+    }
+
+    @Override
+    public Tag getTagById(String tagUuid) throws TagServiceException {
+
+        Tag tag;
+        if (StringUtils.isBlank(tagUuid)) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getCode(),
+                    ErrorMessage.ERROR_CODE_EMPTY_TAG_ID.getMessage());
+        }
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        tag = tagManagementDAO.loadTag(tagUuid, tenantUuid);
+        if (tag == null) {
+            throw new TagServiceClientException(
+                    ErrorMessage.ERROR_CODE_TAG_DATA_DOES_NOT_EXIST.getCode(),
+                    String.format(ErrorMessage.ERROR_CODE_TAG_DATA_DOES_NOT_EXIST.getMessage(), tagUuid));
+        }
+        return tag;
+    }
+
+    @Override
+    public List<String> getTagTypes(int limit, int offset, String sortBy, String sortOrder) throws TagServiceException {
+
+        List<String> tagList;
+        handleSorting(sortBy,sortOrder);
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        tagList = tagManagementDAO.loadTagTypes(limit, offset, sortBy, sortOrder);
+        return tagList;
+    }
+
+    @Override
+    public TagListResult filterTags(int limit, int offset, String sortBy, String sortOrder, String filter)
+            throws TagServiceException {
+
+        TagListResult tagListResult;
+        TagManagementDAO tagManagementDAO = DAOFactory.getInstance().getTagManagementDAO();
+        if (limit < 0) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_LIMIT_VALUE.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_LIMIT_VALUE.getMessage());
+        }
+        if (offset < 0) {
+            throw new TagServiceClientException(ErrorMessage.ERROR_CODE_INVALID_OFFSET_VALUE.getCode(),
+                    ErrorMessage.ERROR_CODE_INVALID_OFFSET_VALUE.getMessage());
+        }
+        handleSorting(sortBy,sortOrder);
+        //TODO do filter validation
```

Comment:
Do we need to fix this? Or is this a future improvement

Reply:
fixed